File: app/src/main/java/com/motrixi/datacollection/fragment/ShowMoreFragment.java
package com.motrixi.datacollection.fragment;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.motrixi.datacollection.DataCollectionActivity;
import com.motrixi.datacollection.content.Contants;
import com.motrixi.datacollection.content.Session;
import com.motrixi.datacollection.utils.DisplayUtil;
import com.motrixi.datacollection.utils.NetworkUtil;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link ShowMoreFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class ShowMoreFragment extends Fragment {

    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;
    private String linkURL;

    private View rootView;
    private DataCollectionActivity parentActivity;

    private RelativeLayout moreLayout;
    private WebView webView;
    private LinearLayout actionBarLayout;
    private TextView tvTitle;

    public ShowMoreFragment() {
        // Required empty public constructor

        parentActivity = DataCollectionActivity.getSharedMainActivity();
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @return A new instance of fragment ShowMoreFragment2.
     */
    // TODO: Rename and change types and number of parameters
    public static ShowMoreFragment newInstance(String param1) {
        ShowMoreFragment fragment = new ShowMoreFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            linkURL = getArguments().getString(ARG_PARAM1);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        //return inflater.inflate(R.layout.fragment_show_more2, container, false);

        initLayout();

        rootView = moreLayout;
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        //parentActivity = (DataCollectionActivity) getActivity();

        //tvTitle.setText(parentActivity.info.value.link_page_title);
        tvTitle.setText(parentActivity.mSession.getLinkTitle());

        customActionBarView();
        initView();
    }

    @SuppressLint("ResourceType")
    private void initLayout() {
        moreLayout = new  RelativeLayout(getActivity());
        RelativeLayout.LayoutParams rootParams = new  RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
        );
        moreLayout.setLayoutParams(rootParams);

        RelativeLayout topLayout = new RelativeLayout(getActivity());
        RelativeLayout.LayoutParams topParams = new  RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                DisplayUtil.dp2px(getActivity(), 50)
        );
        topLayout.setId(Contants.MORE_TOP_ID);
        topLayout.setLayoutParams(topParams);
        topLayout.setBackgroundColor(Color.rgb(0, 150, 182));
        moreLayout.addView(topLayout);

        tvTitle = new TextView(getActivity());
        RelativeLayout.LayoutParams titleParams = new  RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        titleParams.addRule(RelativeLayout.CENTER_IN_PARENT);
        //tvTitle.text = "Detailed Terms"
        tvTitle.setTextSize(20F);
        tvTitle.setTextColor(Color.BLACK);
        tvTitle.setLayoutParams(titleParams);
        topLayout.addView(tvTitle);

        webView = new  WebView(getActivity());
        RelativeLayout.LayoutParams webParams = new  RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
        );
        webParams.addRule(RelativeLayout.BELOW, topLayout.getId());
        webView.setLayoutParams(webParams);
        moreLayout.addView(webView);

    }

    private void initView() {

        initWebView();
        Log.e("link", linkURL);
        webView.loadUrl(linkURL);
        //webView.loadUrl("https://www.motrixi.com/index.php/privacy-policy-2/");

    }

    private void initWebView() {
        WebSettings webSettings = webView.getSettings();

        webSettings.setJavaScriptEnabled(true);
        webSettings.setJavaScriptCanOpenWindowsAutomatically(true);
        webSettings.setAllowFileAccess(true);
        webSettings.setSupportZoom(true);
        webSettings.setJavaScriptEnabled(true);
        webSettings.setDatabaseEnabled(true);
        webSettings.setDomStorageEnabled(true);
        if (NetworkUtil.iConnected(parentActivity)) {
            Log.d("network status", "online");
            webSettings.setCacheMode(WebSettings.LOAD_DEFAULT);
        } else {
            Log.d("network status", "offline");
            webSettings.setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);
        }

        webView.setWebViewClient(webViewClient);
    }

    private void customActionBarView() {

        actionBarLayout = new  LinearLayout(getActivity());
        RelativeLayout.LayoutParams rootParams = new  RelativeLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
        );
        actionBarLayout.setLayoutParams(rootParams);
        actionBarLayout.setGravity(Gravity.CENTER);

        TextView tvTitle = new TextView(getActivity());
        tvTitle.setTextSize(22F);
        //tvTitle.setText(mSession.getConsentDataInfo().value.link_page_title);
        tvTitle.setTextColor(Color.WHITE);
        LinearLayout.LayoutParams titleParams = new  LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        titleParams.rightMargin = DisplayUtil.dp2px(getActivity(), 50);
        tvTitle.setLayoutParams(titleParams);
        tvTitle.setGravity(Gravity.CENTER);

        actionBarLayout.addView(tvTitle);
    }

    private WebViewClient webViewClient = new WebViewClient(){
        @Override
        public void onPageFinished(WebView view, String url) {//页面加载完成
        }

        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {//页面开始加载
        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            Log.i("ansen","url:"+url);

            return super.shouldOverrideUrlLoading(view, url);
        }

    };

}

File: app/src/main/java/com/motrixi/datacollection/fragment/ShowMoreFragment.kt
package com.motrixi.datacollection.fragment

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.Picture
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebView.GONE
import android.webkit.WebView.PictureListener
import android.webkit.WebViewClient
import android.widget.*
import androidx.fragment.app.Fragment
import com.motrixi.datacollection.DataCollectionActivity
import com.motrixi.datacollection.content.Contants
import com.motrixi.datacollection.content.Session
import com.motrixi.datacollection.utils.DisplayUtil
import com.motrixi.datacollection.utils.NetworkUtil
import java.util.function.LongFunction


// TODO: Rename parameter arguments, choose names that match
// the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
private const val ARG_PARAM1 = "param1"
private const val ARG_PARAM2 = "param2"

/**
 * A simple [Fragment] subclass.
 * Use the [ShowMoreFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class ShowMoreFragment : Fragment(), View.OnClickListener {
    // TODO: Rename and change types of parameters
    private var linkURL: String? = null

    private var rootView: View? = null
    private var parentActivity: DataCollectionActivity? = null

    lateinit var moreLayout: RelativeLayout
    lateinit var webView: WebView
    private var actionBarLayout: LinearLayout? = null
    private var mSession: Session? = null
    lateinit var progressBar: ProgressBar


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            linkURL = it.getString(ARG_PARAM1)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        /*rootView = inflater.inflate(R.layout.fragment_show_more, container, false)
        ViewCompat.setElevation(rootView!!, 50f)*/

        initLayout()

        rootView = moreLayout
        return rootView
    }

    @SuppressLint("ResourceType")
    private fun initLayout() {
        moreLayout = RelativeLayout(activity)
        val rootParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        moreLayout.layoutParams = rootParams

        var topLayout = RelativeLayout(activity)
        val topParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            DisplayUtil.dp2px(activity!!, 50)
        )
        topLayout.id = Contants.MORE_TOP_ID
        topLayout.layoutParams = topParams
        topLayout.visibility = View.GONE
        moreLayout.addView(topLayout)

        var ivBack = ImageView(activity)
        val imageParams = RelativeLayout.LayoutParams(
            DisplayUtil.dp2px(activity!!, 25),
            DisplayUtil.dp2px(activity!!, 25)
        )
        imageParams.leftMargin = DisplayUtil.dp2px(activity!!, 20)
        imageParams.addRule(RelativeLayout.CENTER_VERTICAL)
        //ivBack.setImageResource(R.drawable.ic_back)
        ivBack.layoutParams = imageParams
        topLayout.addView(ivBack)

        var tvTitle = TextView(activity)
        val titleParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        titleParams.addRule(RelativeLayout.CENTER_IN_PARENT)
        tvTitle.text = "Detailed Terms"
        tvTitle.textSize = 20F
        tvTitle.setTextColor(Color.BLACK)
        tvTitle.layoutParams = titleParams
        topLayout.addView(tvTitle)


        /*var progressBarLayout = RelativeLayout(activity)
        val barLayoutParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        barLayoutParams.addRule(RelativeLayout.BELOW, topLayout.id)
        progressBarLayout.layoutParams = barLayoutParams
        moreLayout.addView(progressBarLayout)*/

        progressBar = ProgressBar(activity!!)
        val barParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        barParams.addRule(RelativeLayout.CENTER_IN_PARENT)
        progressBar.layoutParams = barParams
        progressBar.isIndeterminate  = true
        moreLayout.addView(progressBar)


        webView = WebView(activity!!)
        val webParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        webParams.addRule(RelativeLayout.BELOW, topLayout.id)
        webView.layoutParams  = webParams
        moreLayout.addView(webView)


        ivBack.setOnClickListener(object : View.OnClickListener{
            override fun onClick(p0: View?) {
                activity!!.onBackPressed()
            }
        })



        //initView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        parentActivity = activity as DataCollectionActivity?
        mSession = Session(activity!!)

        customActionBarView()
        parentActivity!!.actionBar!!.customView = actionBarLayout
        parentActivity!!.actionBar!!.setDisplayShowCustomEnabled(true)

        initView()
    }

    private fun initView() {

        initWebView()
        //webView.loadUrl(Contants.WEB_URL)
//        webView.loadUrl(parentActivity!!.info!!.value!!.terms_link!!)
        Log.e("link", linkURL!!)
        webView.loadUrl(linkURL!!)

        //iv_back.setOnClickListener(this)
    }

    private fun initWebView() {
        var webSettings = webView!!.settings

        webSettings.javaScriptEnabled = true
        webSettings.javaScriptCanOpenWindowsAutomatically = true
        webSettings.allowFileAccess = true
        webSettings.setSupportZoom(true)
        webSettings.javaScriptCanOpenWindowsAutomatically = true
        webSettings.databaseEnabled = true
        webSettings.domStorageEnabled = true
        if (NetworkUtil.iConnected(activity!!)) {
            Log.d("network status", "online")
            webSettings.cacheMode = WebSettings.LOAD_DEFAULT
        } else {
            Log.d("network status", "offline")
            webSettings.cacheMode = WebSettings.LOAD_CACHE_ELSE_NETWORK
        }
        //webSettings.cacheMode = WebSettings.LOAD_NO_CACHE

        webSettings.domStorageEnabled = true
        webSettings.databaseEnabled = true
        //webView.webViewClient = webClient

        webView.setPictureListener(pictureListener)

        webView.webViewClient = object : WebViewClient(){
            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                Log.e("finish", "onPageFinished")
            }

            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                super.onPageStarted(view, url, favicon)
                Log.e("start load", "onPageStarted")
            }

            override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
                return super.shouldOverrideUrlLoading(view, url)
            }


        }

        webView.webChromeClient = object : WebChromeClient(){
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                super.onProgressChanged(view, newProgress)
                if (newProgress == 100) {
                    Log.e("load done", "Finished")
                }
            }
        }
    }

    private fun customActionBarView() {

        actionBarLayout = LinearLayout(activity)
        val rootParams = RelativeLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )

        actionBarLayout!!.layoutParams = rootParams
        actionBarLayout!!.gravity = Gravity.CENTER

        val tvTitle = TextView(activity)
        tvTitle.textSize = 22F
//        tvTitle.text = "Detailed Terms"
        tvTitle.text = mSession!!.consentDataInfo.value!!.link_page_title
        tvTitle.setTextColor(Color.WHITE)
        val titleParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        titleParams.rightMargin = DisplayUtil.dp2px(activity!!, 50)
        tvTitle.layoutParams = titleParams
//        tvCancel.setBackgroundColor(Color.rgb(0, 150, 182))  //#0096B6
        tvTitle.gravity = Gravity.CENTER

        actionBarLayout!!.addView(tvTitle)
    }


    private val pictureListener : PictureListener = object : PictureListener {
        override fun onNewPicture(p0: WebView?, p1: Picture?) {
            Log.e("picture", "listener")
            if (progressBar != null) {
                if (progressBar.isShown) {
                    progressBar.visibility = View.GONE
                }
            }
        }

    }

    private val webClient = object : WebViewClient() {

        override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
            super.onPageStarted(view, url, favicon)
            Log.e("start load", "onPageStarted")

        }

        override fun onPageFinished(view: WebView?, url: String?) {
            super.onPageFinished(view, url)
            Log.e("finish", "onPageFinished")
        }

        override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
            Log.d(TAG, url!!)
            return false
        }


    }

    override fun onClick(view: View?) {
        /*when (view!!.id) {
            R.id.iv_back -> {
                activity!!.onBackPressed()
            }
            else -> {
            }
        }*/
    }

    companion object {

        val TAG=ShowMoreFragment::class.simpleName  //定义Log的TAG

        @JvmStatic
        fun newInstance(param1: String) =
            ShowMoreFragment().apply {
                arguments = Bundle().apply {
                    putString(ARG_PARAM1, param1)
                }
            }
    }


}